http/README.md

```diff
@@ -0,0 +1,114 @@
+
+Authors: [@marten-seemann, @MarcoPolo]
+
+Interest Group: [@lidel]
+
+[@marten-seemann]: https://github.com/marten-seemann
+[@MarcoPolo]: https://github.com/MarcoPolo
+[@lidel]: https://github.com/lidel
+
+## Introduction
+
+This document defines how libp2p nodes can offer and use an HTTP transport alongside their other transports to support application protocols with HTTP semantics. This allows a wider variety of nodes to participate in the libp2p network, for example:
+
+- Browsers communicating with other libp2p nodes without needing a WebSocket, WebTransport, or WebRTC connection.
+- HTTP only edge workers can run application protocols and respond to peers on the network.
+- `curl` from the command line can make requests to other libp2p nodes.
+
+As well as allowing application protocols to make use of HTTP intermediaries such as HTTP caching and layer 7 proxying and load balancing. This is all in addition to the existing features that libp2p provides such as:
+
+- Connectivity – Work on top of WebRTC, WebTransport, QUIC, TCP, or an HTTP transport.
+- Hole punching – Work with peers behind NATs.
+- Peer ID Authentication – Authenticate your peer by their libp2p peer id.
+- Peer discovery – Learn about a peer given their peer id.
+
+## HTTP Transport vs HTTP Semantics
+
+HTTP is a bit of an overloaded term. This section aims to clarify what we’re talking about when we say “HTTP”.
+
+*HTTP semantics* ([RFC 9110](https://www.rfc-editor.org/rfc/rfc9110.html)) is the stateless application-level protocol that you work with when writing HTTP apis (for example).
+
+*HTTP transport* is the thing that takes your high level request/response defined in terms of HTTP semantics and encodes it and sends it over the wire.
+
+When this document says *HTTP* it is generally referring to *HTTP semantics*.
+
+## Interoperability with existing HTTP systems
+
+A goal of this spec is to allow libp2p to be able to interoperate with existing HTTP servers and clients. Care is taken in this document to not introduce anything that would break interoperability with existing systems.
+
+## HTTP Transport
+
+Nodes MUST use HTTPS (i.e. they MUST NOT use plaintext HTTP). It is RECOMMENDED to use HTTP/2 and HTTP/3.
+
+Nodes signal support for their HTTP transport using the `/http` component in their multiaddr. e.g. `/dns4/example.com/tls/http` . See the [HTTP multiaddr component spec](https://github.com/libp2p/specs/pull/550) for more details.
+
+## Namespace
+
+libp2p does not squat the global namespace. libp2p application protocols can be discovered by the [well-known resource](https://www.rfc-editor.org/rfc/rfc8615) `.well-known/libp2p`. This allows server operators to dynamically change the URLs of the application protocols offered, and not hard-code any assumptions how a certain resource is meant to be interpreted.
+
+```json
+
+{
+    "services": {
+        "/kad/1.0.0": "/kademlia/",
+        "/ipfs-http/1.0.0": "/",
```

Comment:
Some more thoughts on this:

1. The path alone is not enough to form a request. We need more (out-of-band) information like what method to use.
2. Do you expect a protocol-spec like kademlia to be extended on how it can be accessed from an HTTP transport? (And which methods need authentication for example?)

Reply:
I'm a bit hesitant to add the ability to put more metadata without a use case behind it, but making it a map with path seems easy enough, and would allow adding new fields in a backwards compatible way, which is nice.

---

> 1. The path alone is not enough to form a request. We need more (out-of-band) information like what method to use.

This is up to the application protocol. The application protocol defines how it works and what HTTP methods it uses for what. This metadata only describes where the application protocol is mounted at. It doesn't describe the application protocol.

> 2. Do you expect a protocol-spec like kademlia to be extended on how it can be accessed from an HTTP transport? (And which methods need authentication for example?)

Yes. That's up to the application protocol. Kademlia could make use of the simple request-response abstraction in #561 to define this, but ultimately the application protocol decides tis.

Reply:
I've updated this to be a map from protocol name to metadata that includes a path.

Reply:
> > 3\. Do you expect a protocol-spec like kademlia to be extended on how it can be accessed from an HTTP transport? (And which methods need authentication for example?)
> 
> Yes. That's up to the application protocol. Kademlia could make use of the simple request-response abstraction in #561 to define this, but ultimately the application protocol decides tis.

I see, thanks for explaining. Can we make this clearer? Something like:

This mapping only defines a URL namespace for certain applications protocols. It is entirely up to the application protocol (like kademlia) to define how it can be interacted with over HTTP.